### ISSUE TITLE
Possible problem with new users and Slurm associations

### ISSUE THREAD

**Issue by verdurin** [ComputeCanada/puppet-magic_castle#106]
We've been using `Mokey` for users to add themselves to a cluster.
For the first cohort of users, adding them to the relevant group ensured that they had a Slurm association and could submit jobs.

However, it appears that users added later do not pick up the Slurm association even once they have been added to the correct group. 

Wondering whether this is similar to the other issue caused by the rotation of log files?

**Comment by cmd-ntrf** [comment]
Could be, yes.

What was the timelapse between the two cohorts?

Was Slurm account association the only thing missing ? Home, projects and scratch were correctly created?

**Comment by verdurin** [comment]
The timelapse is roughly 24 hours or more.

`/home` and `/scratch` were created, but not `/projects`.

**Comment by cmd-ntrf** [comment]
This was with MC 10.0?

**Comment by verdurin** [comment]
No, `9.3` up to commit `a4ee76a5`.

Thought I'd pulled in updates after that, but it appears not.

**Comment by cmd-ntrf** [comment]
Can you look into the project creation daemon see if there are any errors?

On mgmt1
```
sudo systemctl status mkproject
```

```
sudo journalctl -u mkproject
```

**Comment by cmd-ntrf** [comment]
> No, `9.3` up to commit `a4ee76a5`.

This is a commit for `magic_castle`. Have you cherry-picked https://github.com/ComputeCanada/puppet-magic_castle/commit/13f5604c4c9b6bec26a458ebfaa5fe6b4ea7414f ?

**Comment by verdurin** [comment]
On Mon, Mar 01, 2021 at 06:49:02AM -0800, Félix-Antoine Fortin wrote:
> > No, `9.3` up to commit `a4ee76a5`.
> 
> This is a commit for `magic_castle`. Have you cherry-picked https://github.com/ComputeCanada/puppet-magic_castle/commit/13f5604c4c9b6bec26a458ebfaa5fe6b4ea7414f ?

Yes, already have those fixes.

**Comment by verdurin** [comment]
On Mon, Mar 01, 2021 at 06:44:22AM -0800, Félix-Antoine Fortin wrote:
> Can you look into the project creation daemon see if there are any errors?
> 
> On mgmt1
> ```
> sudo systemctl status mkproject
> ```
> 
> ```
> sudo journalctl -u mkproject
> ```
> 

The service is running, and I can't see any obvious errors.

No users created after 26th February have received Slurm associations
automatically.

**Comment by cmd-ntrf** [comment]
Some more debugging hints:
- Does the name of the group that was associated with the users with FreeIPA starts with one of the following prefix : `def`, `ctf`, `rpp`, or `rrg`? This is the Compute Canada convention to identify POSIX groups that are associated with Slurm accounts. Other groups are ignored by the daemon.
- Can you confirm line 11 of `/sbin/mkproject.sh` reads like this:
```tail -F /var/log/dirsrv/slapd-*/access |```
- Can you confirm that every 24 hours in  `journalctl -u mkproject`, there is a mention the log file has moved?

**Comment by verdurin** [comment]
On Tue, Mar 02, 2021 at 07:51:33AM -0800, Félix-Antoine Fortin wrote:
> Some more debugging hints:
> - Does the name of the group that was associated with the users with FreeIPA starts with one of the following prefix : `def`, `ctf`, `rpp`, or `rrg`? This is the Compute Canada convention to identify POSIX groups that are associated with Slurm accounts. Other groups are ignored by the daemon.

Yes, I just used the default `def-sponsor00` group.

> - Can you confirm line 11 of `/sbin/mkproject.sh` reads like this:
> ```tail -F /var/log/dirsrv/slapd-*/access |```

Yes:
```tail -F /var/log/dirsrv/slapd-*/access |```

> - Can you confirm that every 24 hours in  `journalctl -u mkproject`, there is a mention the log file has moved?
> 

Yes, there's the usual "following new file" message.

**Comment by cmd-ntrf** [comment]
I suspect the issue is with the two counters in `mkproject.sh`:
https://github.com/ComputeCanada/puppet-magic_castle/blob/master/site/profile/files/accounts/mkproject.sh#L6

Each time a user is added to a group, there is a line corresponding to this regex:
```
MOD dn=\"cn=\K((def|rrg|ctg|rpp)-[a-z0-9A-Z_-]*)(?=,cn=groups)"
```
that appears in /var/log/dirsrv/slapd-*/access

The line however does not mention which users has been added to the group. So we use an ldapsearch to identify which users are part of the group. The ldapsearch results may contain more usernames than the one recently added because by the time the line appeared in the slapd log, more users can have been added to a group.

So to avoid trying to add users that would already be part of a Slurm account, I tried to enumerate the number of users and the number of time `mkproject.sh` saw the preceding regex in the log. But it is possible the count goes out of control for reasons left to determine, and the Slurm account appending is skipped because `mkproject.sh` thinks the user is already part of it.

There are few things that can go wrong with the count logic, so I'll try to get rid of it.

**Comment by cmd-ntrf** [comment]
I have a new version of mkproject.sh that addresses the issues raised here:
https://github.com/ComputeCanada/puppet-magic_castle/blob/mkproject_redo/site/profile/files/accounts/mkproject.sh

It worked on most situations I have shoot at it so far. It should be part of MC 10.1.

**Comment by cmd-ntrf** [comment]
This should now be fixed in MC 10.2.
